go/vt/vtgate/engine/semi_join.go

```diff
@@ -62,24 +63,26 @@ func (jn *SemiJoin) TryExecute(ctx context.Context, vcursor VCursor, bindVars ma
 
 // TryStreamExecute performs a streaming exec.
 func (jn *SemiJoin) TryStreamExecute(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool, callback func(*sqltypes.Result) error) error {
-	joinVars := make(map[string]*querypb.BindVariable)
 	err := vcursor.StreamExecutePrimitive(ctx, jn.Left, bindVars, wantfields, func(lresult *sqltypes.Result) error {
+		joinVars := make(map[string]*querypb.BindVariable)
 		result := &sqltypes.Result{Fields: lresult.Fields}
 		for _, lrow := range lresult.Rows {
 			for k, col := range jn.Vars {
 				joinVars[k] = sqltypes.ValueBindVariable(lrow[col])
 			}
-			rowAdded := false
+			var rowAdded atomic.Bool
 			err := vcursor.StreamExecutePrimitive(ctx, jn.Right, combineVars(bindVars, joinVars), false, func(rresult *sqltypes.Result) error {
-				if len(rresult.Rows) > 0 && !rowAdded {
-					result.Rows = append(result.Rows, lrow)
-					rowAdded = true
+				if len(rresult.Rows) > 0 {
+					rowAdded.Store(true)
 				}
 				return nil
 			})
 			if err != nil {
 				return err
 			}
+			if rowAdded.Load() {
```

Comment:
Why is this one atomic? If this needs an atomic since it otherwise races, I imagine that the `append` on the next line also races and needs to be guarded with a lock then? 

If this doesn't race, it doesn't need to be atomic?

Reply:
Yes i know, I made the changes such that i could avoid using this atomic thing, but in the callback right above we are writing to it. We are only setting it to true but golang still complains in a `-race` test saying there is a concurrent write. So I had to make it atomic to avoid that. That being said, for strictly correctness purposes, we didn't need to make it atomic, but I couldn't figure out how to make the test work without making it atomic.

Reply:
both are racy as the internal `StreamExecute` can do parallel execute to multiple shards

Reply:
@harshit-gangal Right, so that means we need to guard the `append` too right? Or otherwise we lose data there potentially?

Reply:
No we don't. The Result field is local to each outer StreamExecute call. Its initialized in the callback.

Reply:
results is local to outer callback but append happens inside the second callback, which itself can execute in parallel causes race in append.

Reply:
If you really want to test that, you would need a test that does scatter call for the inside stream execute (right side of semi-join)

Reply:
Append is not inside the second callback. That is the entire point.

Reply:
The test I added does a scatter for the inside semi-join and the outside semi-join too.

Reply:
This looks good.